slowquant/qiskit_interface/wavefunction.py

```diff
@@ -293,19 +314,38 @@ def change_primitive(self, primitive: BaseEstimator | BaseSampler, verbose: bool
         self.QI.total_device_calls = 0
         self.QI.total_shots_used = 0
         self.QI.total_paulis_evaluated = 0
-        self.QI.cliques = Clique()
+        self.QI._reset_cliques(verbose=verbose)  # pylint: disable=protected-access
         self.QI._Minv = None  # pylint: disable=protected-access
-        self.QI._primitive = primitive  # pylint: disable=protected-access
-        # IMPORTANT: Shot number in primitive gets always overwritten if a shot number is defined in QI!
-        if self.QI.shots is not None and verbose:
-            print(
-                "Number of shots defined in new primitive are ignored as there is a number defined in the QI of ",
-                self.QI.shots,
-            )
-            print("If you want to change the number of shots, do this manually.")
-            print("Set the number of shots manually to None if you run an ideal simulator.")
-        self.QI.shots = self.QI.shots  # Redo shot check with new primitive
-        self.QI.ISA = self.QI.ISA  # Redo ISA parameter check
+
+        # Reset circuit and initiate re-transpiling
+        ISA_old = self.QI.ISA
+        self._reconstruct_circuit()  # Reconstruct circuit but keeping parameters
+        self.QI.ISA = ISA_old  # Redo ISA including transpilation if requested
+        self.QI.shots = self.QI.shots  # Redo shots parameter check
```

Comment:
Uhm?

Reply:
See above. We need to check ISA and shots again when we change primitive because. for example, we did ideal simulator (shots = None) and now we change to SamplerV2 that needs to have shots defined as it cannot do ideal simulations. So we run through the shot setter again where all this is ensured. 
The alterantive would be to copy paste the setter routine for ISA, shots, etc. into the change_primitive and do all the possible changes needed hard coded into the change_primitive function. The current way looks odd but is efficient (and smart?)